File: frontend/src/types/purchaseOrder.ts
import type { PurchaseOrderItem } from "./purchaseOrderItem";
import type { OrderStatus } from "./OrderStatus";

export interface PurchaseOrder {
  idPurchaseOrder: number;
  idCompany: number;
  idSupplier: number;
  supplierName: string;
  orderNumber: string;
  status: OrderStatus;
  arcNumber?: string | null;
  orderedAt: string;
  expectedDeliveryDate: string;
  receivedAt?: string | null;
  totalAmount: number;
  createdAt: string;
  isAutoRecommended: boolean;
  items?: PurchaseOrderItem[];
}

export interface PurchaseOrderCreate {
  idCompany: number;
  idSupplier: number;
  orderNumber: string;
  status: OrderStatus;
  orderedAt?: string;
  expectedDeliveryDate?: string;
  totalAmount?: number;
  isAutoRecommended?: boolean;
}

export interface PurchaseOrderUpdate {
  status: OrderStatus;
  expectedDeliveryDate?: string;
  receivedAt?: string | null;
  totalAmount?: number;
}

export interface PurchaseOrderItemReceiveDTO {
  idPurchaseOrderItem: number;
  quantityReceivedNow: number;
  unitPrice?: number;
}

export interface PurchaseOrderReceiveDTO {
  items: PurchaseOrderItemReceiveDTO[];
}

export interface PurchaseOrderArcUpdateDTO {
  arcNumber: string;
  expectedDeliveryDate: string;
}

File: backend/buylogic/src/main/java/com/buylogic/model/PurchaseOrder.java
package com.buylogic.model;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;

import com.buylogic.model.enums.OrderStatus;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Entity
@Table(name = "purchase_order")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class PurchaseOrder {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer idPurchaseOrder;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "id_company", nullable = false)
    private Company company;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "id_supplier", nullable = false)
    private Supplier supplier;

    @Column(name = "order_number", length = 100)
    private String orderNumber;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 30)
    private OrderStatus status = OrderStatus.DRAFT;

    private LocalDateTime orderedAt;

    @Column(name = "arc_number", length = 100)
    private String arcNumber;

    private LocalDate expectedDeliveryDate;

    private LocalDateTime receivedAt;

    @Column(name = "total_amount", precision = 14, scale = 2)
    private BigDecimal totalAmount;

    @Column(nullable = false)
    private LocalDateTime createdAt;

    @Column(nullable = false)
    private LocalDateTime updatedAt;

    @Column(name = "is_auto_recommended", nullable = false)
    private Boolean isAutoRecommended = false;

    @OneToMany(mappedBy = "purchaseOrder")
    private List<PurchaseOrderItem> items;

    @PrePersist
    protected void onCreate() {
        LocalDateTime now = LocalDateTime.now();
        createdAt = now;
        updatedAt = now;
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}
